src/firewall.ts

```diff
@@ -460,147 +469,342 @@ class WirefilterFirewallRule implements FirewallRule {
+            id: rule.id,
+            expression: new WirefilterFirewallRule(this.wirefilter, this.scheme, rule.expression),
+            action: rule.action,
+        });
+        return this;
+    }
+
+    /**
+     * Iterates through the list of rules and returns the matching rule id
+     *
+     * @param request incoming request to be matched against
+     * @returns the provided rule id if a successful match is made, undefined otherwise
+     */
+    matchRequest(request: Request): RequestMatchResults {
+        const results = new RulesetMatchResults();
+        const execCtx = ExecutionContext.buildFromRequest(this.wirefilter, this.scheme, this.defaultLists, request);
+        try {
+            for (const rule of this.rules) {
+                if (!rule.expression.matchUsingContext(execCtx)) continue;
+
+                results.matches.push({
+                    ruleId: rule.id,
+                    action: { ...rule.action }
+                });
+
+                if (isTerminalAction(rule.action)) {
+                    break;
+                }
+                if (rule.action.type === 'skip' && rule.action.ruleset === 'current') {
+                    break;
+                }
+            }
+        } finally {
+            execCtx.free();
+        }
+        return results;
+    }
+}
+
+export interface RuleMatch {
+    ruleId: string
+    action: FirewallAction
+}
+
+export interface TerminatingActionMatch {
+    ruleId: string
+    action: TerminalAction
+}
+
+const isTerminatingActionMatch = (match: RuleMatch): match is TerminatingActionMatch => {
+    return isTerminalAction(match.action);
+};
+
+export interface RequestMatchResults {
+    matches: RuleMatch[]
+    skippedPhases: Phase[]
+    skippedProducts: Product[]
+    loggedRules: string[]
+    terminatedEarly: boolean
+    terminalAction: { action: { type: 'no_match' } } | TerminatingActionMatch
+}
+
+class RulesetMatchResults implements RequestMatchResults {
+    matches: RuleMatch[] = [];
+
+    get skippedPhases(): Phase[] {
+        const phases = new Set<Phase>();
+        for (const m of this.matches) {
+            if (m.action.type === 'skip') {
+                m.action.phases?.forEach(p => phases.add(p));
+            }
+        }
+        return Array.from(phases);
+    }
+
+    get skippedProducts(): Product[] {
+        const products = new Set<Product>();
+        for (const m of this.matches) {
+            if (m.action.type === 'skip') {
+                m.action.products?.forEach(p => products.add(p));
+            }
+        }
+        return Array.from(products);
+    }
+
+    get loggedRules(): string[] {
+        return this.matches.filter(m => m.action.type === 'log').map(m => m.ruleId);
+    }
+
+    get terminatedEarly(): boolean {
+        return this.matches.some(m => m.action.type === 'skip' && m.action.ruleset === 'current');
+    }
+
+    get terminalAction(): TerminatingActionMatch | { action: { type: 'no_match'; } } {
```

Comment:
similar to above think this should also return the skip + current combo if that was the terminating match